### PR TITLE
B2C cache fix

### PR DESCRIPTION
**Pull request by trwalke**
Injecting tenantID  into the returned B2C ID Token to enable caching for b2c scenarios.

Removed check for preferedUserName from MSAL caching on B2C scenarios.

VSTS bug 533786

This fix was tested E2E with the MSAL desktop App. Both interactive and silent B2C calls were tested

### CODE REVIEW

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: I'd wrap this up in a private method, smth like "InjectTenantIdB2C" or smth and add more details about B2C behaviour #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: No unit tests need to change? #Closed

**Review by bgavrilMS — COMMENTED** [review]
LGTM, would be great if we can add a unit test
#Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >//PreferredUsername is not required in B2C cache scenarios [](start = 12, length = 58)

Did you test this? Can you aquire a token silently? #Closed

**Comment by henrik-me** [comment]
This is not injection of tenant id. this is using tenant id from the request parameters in case no tenant id is returned. pls. rename the method which does this. (the method you will have when having done the refactoring suggested by Bogdan)

#Closed #Closed

**Comment by henrik-me** [comment]
Please clearly state that you have tested silent scenarios for B2C with this change and that everything works as expected.  #Closed

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (oldalton) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: it's probably me reading the code wrong, but why do we pass the "tenantId" from request params authority to the MsalAccessTokenCacheItem? #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (oldalton) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: Unless MSAL .NET uses preferred username for its legacy caching purposes, it shouldn't be required for caching at all.  #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: Yes
#Closed #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: The MsalAccessTokenCacheItem represents a data entry within the token cache. This item has a function called GetKey() which will generate a key for the token cache dictionary to use with the tenantID and other data within the MsalAccessTokenCacheItem. 
see: [MsalAccessTokenCacheItem Line 114](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/a25be22e9d667050b96eb4517f1401c2dd97a119/core/src/Cache/MsalAccessTokenCacheItem.cs#L114), [TokenCacheAccessor Line 57](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/a25be22e9d667050b96eb4517f1401c2dd97a119/core/src/Features/Cache/TokenCacheAccessor.cs#L57)
 #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: From my testing, when I create a token cache entry, the preferredUserName is not present in either the key or the value. However, this check will prevent a response from being cached if the preferedUserName is not set regardless. I am not sure if I am missing something.

@SomkaPe can you explain why there is a check for preferedUserName? #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: #Closed #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: #Closed #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @trwalke did you test this and other flows with-out the preferredUserName set?  #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: Move to XML comments #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >if (requestParams.Authority.AuthorityType == Core.Instance.AuthorityType.B2C) [](start = 12, length = 77)

invert if, it will reduce nesting and thus improve readability #Closed

File: msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs
Comment (henrik-me) [inline msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs:905]: What sets the authority type to B2C? Suggestion add an assert which validates that Authority Type is B2C (then you know the new code is run)

**Comment by henrik-me** [comment]
Waiting for confirmation in writing that the scenarios has been run. Please point to test cases which contains this validation.

---
In reply to: [417670956](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#issuecomment-417670956) [](ancestors = 417670956)

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (oldalton) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: Yes, that makes sense. However I was referring to where the "tenantId" is coming from:

`var tenantId = Authority.CreateAuthority(requestParams.TenantUpdatedCanonicalAuthority, false).GetTenantId();`

To me it seems that it's coming from developer requested authority and not idToken.TenantId.  #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (oldalton) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: PreferredUsername has to be present in the Account object in the payload. But it's not used as part of the cache key. Therefore, it's not crucial for the caching purposes.  #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >string.IsNullOrEmpty(idToken.TenantId) [](start = 20, length = 38)

also you can move this up to be the first check made before checking for autorityType. If tenantid exists in the idtoken, nothing more is needed. This also means the method should probably be renamed to e.g.: EnsureTenantIDIsSpecifiedForB2CAuthority #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @henrik-me yes, I tried this flow and the standard authorization code flow. both failed to seed the cache if the preferedUserName was not present even though the preferedUserName is not used for caching purposes. #Closed

File: msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs
Comment (trwalke) [reply msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs:905]: @henrik-me The authority type is set internally by the Authority class. It checks the authority for different formats and determines the authority type from that. 
B2C authority's take the form of https://login.microsoftonline.com/tfp/someTenantID/somePolicy/oauth2/v2.0/authorize

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @oldalton The requestParams.TenantUpdatedCanonicalAuthority is set by the id token. what is happening here is the requestParams is updated with the new authority returned from the idToken in case the user uses the common authority for authentication. 

Thus, https://login.microsoftonline.com/common/ will be replaced with https://login.microsoftonline.com/SomeTenantID/... in the requestParams.TenantUpdatedCanonicalAuthority property

[See RequestBase.cs Line 228](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/e38e4073e9cc6b3c986feb0572286ab66dbb4202/msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs#L228) #Closed

Comment (oldalton) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: That's a bit confusing, but thanks for looking it up :)
Can we just pass idToken.tenantId into cache item? Looking only at this code, it seems like we're trying to get tenantId from the idToken and falling back to tenantId from requested authority if it's not present. So, it's a bit confusing that we're not passing the result idToken.tenantId into cache item itself. #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (oldalton) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: What exactly failed in the cache? #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: remove. otherwise you are not reducing nesting. Also it's not required because in the other case you have already returned #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: Did you test with this change or before this change? What happens with this change. Did you validate where else the preferredUserName is used? And why?

---
In reply to: [215019267](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215019267) [](ancestors = 215019267)

**Comment by trwalke** [comment]
This fix was tested E2E with the MSAL desktop App. Both interactive and silent B2C calls were tested. 
1st, a token was acquired interactivly for B2C. Then, the access token was expired and the refresh token in the cache was used silently to acquire a new access token.

For the preferedUserName testing, I acquired a token using the standard authorization code flow, put a break point on the location in code where the successful token response is returned from the STS and set the preferedUserName to empty. MSAL skipped the caching of the token because the preferedUserName  was not set. This is expected behavior.

This same test was performed with the B2C flow instead and the token was cached. This is expected behavior because the B2C scenario does not always return a preferedUserName at the moment

@henrik-me VSTS is still down, ill add a link to the test case when service is restored #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @oldalton The application fails to put the access token and refresh token into the cache when the id token had an empty PreferedUserName field. 

@henrik-me this was tested before and after my change. Whenever I set the preferredUserName to empty, the access tokens and refresh tokens were not put into the cache except for the B2C scenario

as for what the preferredUserName is used for, it is used in the iAccount objects and it is used in the TokenCacheNotificationArgs parameters that are used in the BeforeWrite, BeforeAccess and AfterAccess delegates, giving the user the ability to use the preferredUserName as they see fit #Closed

Comment (oldalton) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: But access tokens and refresh tokens weren't written to cache because of this if condition, right? Or you removed this if conditions and tested then? #Closed

Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: Yes, this condition was removed specifically from the B2C case so that tokens are cached #Closed

File: msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs
Comment (henrik-me) [reply msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs:905]: Great. then it should be no issue to validate that the B2C AuthorityType is set.

---
In reply to: [215021421](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215021421) [](ancestors = 215021421)

**Comment by henrik-me** [comment]
Approving (after confirmation offline). However should not be merged until a build and automated tests have passed. Including a final statement of which manual tests have been applied to validate the fix and the tests completed for regression testing purposes. Also still one minor utest ask. #Closed

**Review by henrik-me — APPROVED** [review]
:shipit:

**Comment by SomkaPe** [comment]
:shipit:  #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: >Not returned from server [](start = 54, length = 24)

suggestion: preferred_username not in idtoken (also rename the variable) #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs:129]: is this a TryParse ? or could it throw an exception? #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/TokenCache.cs]: > var preferredUsername = !String.IsNullOrWhiteSpace(idToken?.PreferredUsername)? idToken.PreferredUsername : notReturnedFromServer; [](start = 11, length = 131)

note, the idea of the named method was to make it clear why we did this change. I suppose now we are back to having to add a comment somewhere...? The information will be lost that we have done this to mitigate the B2C scenario... even though the code will ofcourse work for non B2C and thus will be simpler...

I suppose I suggest adding a comment that this is the typical case for B2C

File: msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs
Comment (henrik-me) [reply msal/tests/Test.MSAL.NET.Unit/CacheTests/TokenCacheTests.cs:905]: Would still have linked the assert on the authority.type ... though of little value... also I suppose you latest changes would also work for missing PreferredUserName for non B2C flows. I suggest you add that test case as well

---
In reply to: [215088757](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215088757) [](ancestors = 215088757,215021421)

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @henrik-me Yes, however, our intention here was to design a solution that will resolve all issues with the preferred username so that even when the B2C releases a fix for this, the code will work without us needing to update it. Thus, making this a more generic solution for all issues that could potentially result in a missing tenantid or prefered user name. I think a more generic comment about the necessity of a non-null preferred user name will suffice. thoughts? #Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs:129]: It does wrap the parse operation in a try block. It also checks for null on the id token.
#Closed

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: #Closed #Closed

**Comment by trwalke** [comment]
Cache scenarios have been tested to verify that the cache sharing between 
- ADAL V3 and ADAL V4
- MSAL V2 and ADAL V4

is working

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: I was not explaining my self clearly enough. With the change I proposed it will still work with the update from B2C, however what you have removed now is the need to have preferred username specified for all other scenarios as well. Did you validate that this is all good? Are you sure we will not run into issues (at least usability issues) ?

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (henrik-me) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs:129]: the convention for this is to name the method TryParse ... this improves readability and avoids me having to ask or go research. pls. update the name

---
In reply to: [215472647](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#discussion_r215472647) [](ancestors = 215472647)

**Comment by henrik-me** [comment]
If this works we have introduced AI... I believe you might have a typo here. The first line should probably have been ADAL V3 and ADAL V4 

---
In reply to: [419271270](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#issuecomment-419271270) [](ancestors = 419271270)

**Comment by henrik-me** [comment]
Seems like you did not address the utest ask I had.

File: msal/src/Microsoft.Identity.Client/TokenCache.cs
Comment (trwalke) [reply msal/src/Microsoft.Identity.Client/TokenCache.cs]: @henrik-me Yes, these changes were made with the guidance of @SomkaPe. The removal of the prefered_userName requirement will allow the cache to store tokens with this field missing. However, the token lookup operations do not depend on the preferred username so there will be no change in the behavior of cache lookup operations. I ran this new change through our caching scenarios and no issues were found. We can follow up tomorrow during stand up.

**Comment by trwalke** [comment]
@henrik-me  Yes, you are correct. I updated the comment

in reply to: [419302357](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#issuecomment-419302357)

**Comment by trwalke** [comment]
@henrik-me  I believe, the test was to confirm that the token cache did not cache responses that did not have the prefered_username available in the id token. With the prefered_username requirement gone, this is no longer a test we can make. 

Also, the new B2c cache test will cover scenarios where an idtoken is missing the prefered_username. I think I may rename that test to be more generalized.

in reply to: [419302657](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1241#issuecomment-419302657)

**Pull request merged**
